pipeline_server/src/main.rs

```diff
@@ -0,0 +1,873 @@
+            let json_string = serde_json::to_string(&UpdateProjectResponse { version }).unwrap();
+            HttpResponse::Ok()
+                .insert_header(CacheControl(vec![CacheDirective::NoCache]))
+                .content_type(mime::APPLICATION_JSON)
+                .body(json_string)
+        })
+        .unwrap_or_else(|e| http_resp_from_error(&e))
+}
+
+/// `/compile_project` request parameters.
+#[derive(Deserialize)]
+struct CompileProjectRequest {
+    /// Project id.
+    project_id: ProjectId,
+    /// Latest project version known to the client.
+    version: Version,
+}
+
+/// Queue project for compilation.
+///
+/// The client should poll the `/project_status` endpoint
+/// for compilation results.
+///
+/// # HTTP errors
+///
+/// * `NOT_FOUND` - `project_id` does not exist in the database.
+/// * `CONFLICT` - project version specified in the request doesn't match the
+///   latest project version in the database.
+#[post("/compile_project")]
+async fn compile_project(
+    state: WebData<ServerState>,
+    request: web::Json<CompileProjectRequest>,
+) -> impl Responder {
+    state
+        .db
+        .lock()
+        .await
+        .set_project_pending(request.project_id, request.version)
+        .await
+        .map(|_| HttpResponse::Ok().finish())
+        .unwrap_or_else(|e| http_resp_from_error(&e))
+}
+
+/// `/cancel_project` request parameters.
+#[derive(Deserialize)]
+struct CancelProjectRequest {
+    /// Project id.
+    project_id: ProjectId,
+    /// Latest project version known to the client.
+    version: Version,
+}
+
+/// Cancel outstanding compilation request.
+///
+/// The client should poll the `/project_status` endpoint
+/// to determine when the cancelation request completes.
+///
+/// # HTTP errors
+///
+/// * `NOT_FOUND` - `project_id` does not exist in the database.
+///
+/// * `CONFLICT` - project version specified in the request doesn't match the
+///   latest project version in the database.
+#[post("/cancel_project")]
+async fn cancel_project(
+    state: WebData<ServerState>,
+    request: web::Json<CancelProjectRequest>,
+) -> impl Responder {
+    state
+        .db
+        .lock()
+        .await
+        .cancel_project(request.project_id, request.version)
+        .await
+        .map(|_| HttpResponse::Ok().finish())
+        .unwrap_or_else(|e| http_resp_from_error(&e))
+}
+
+/// `/delete_project` request parameters.
+#[derive(Deserialize)]
+struct DeleteProjectRequest {
+    project_id: ProjectId,
+}
+
+/// Delete a project.
+///
+/// # HTTP errors
+///
+/// * `BAD_REQUEST` - the project has one or more running pipelines.
+///
+/// * `NOT_FOUND` - `project_id` does not exist in the database.
+#[post("/delete_project")]
+async fn delete_project(
+    state: WebData<ServerState>,
+    request: web::Json<DeleteProjectRequest>,
+) -> impl Responder {
+    let db = state.db.lock().await;
+
+    match db.list_project_pipelines(request.project_id).await {
+        Ok(pipelines) => {
```

Comment:
I am not sure that the logic of this should be in the web server. 
the web server is just an API on top of some abstract controller.

Reply:
I think the planned switch to OpenAPI will take care of this.